core/src/main/java/org/commonjava/maven/galley/io/SpecialPathManagerImpl.java

```diff
@@ -15,61 +15,129 @@
 import org.commonjava.maven.galley.model.ConcreteResource;
-import org.commonjava.maven.galley.model.FilePatternMatcher;
 import org.commonjava.maven.galley.model.Location;
-import org.commonjava.maven.galley.model.PathPatternMatcher;
 import org.commonjava.maven.galley.model.SpecialPathInfo;
-import org.commonjava.maven.galley.model.SpecialPathMatcher;
 import org.commonjava.maven.galley.model.Transfer;
 import org.commonjava.maven.galley.spi.io.SpecialPathManager;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import javax.annotation.PostConstruct;
 import javax.enterprise.context.ApplicationScoped;
 import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Comparator;
+import java.util.Collection;
 import java.util.List;
 import java.util.Map;
-import java.util.TreeMap;
 import java.util.concurrent.ConcurrentHashMap;
 
+import static org.commonjava.maven.galley.io.SpecialPathConstants.PKG_TYPE_MAVEN;
+import static org.commonjava.maven.galley.io.SpecialPathConstants.MVN_SP_PATH_SET;
+
 /**
  * Created by jdcasey on 1/27/16.
  */
 @ApplicationScoped
 public class SpecialPathManagerImpl
         implements SpecialPathManager
 {
-    private List<SpecialPathInfo> specialPaths;
+    private static final Logger logger = LoggerFactory.getLogger( SpecialPathManagerImpl.class );
+
+    private List<SpecialPathInfo> stdSpecialPaths;
+
+    private Map<String, SpecialPathSet> pkgtypes;
 
     public SpecialPathManagerImpl()
     {
-        specialPaths = new ArrayList<>();
-        specialPaths.addAll( SpecialPathConstants.STANDARD_SPECIAL_PATHS );
+        initPkgPathSets();
+    }
+
+    @PostConstruct
+    public void initPkgPathSets()
+    {
+        stdSpecialPaths = new ArrayList<>();
+        stdSpecialPaths.addAll( SpecialPathConstants.STANDARD_SPECIAL_PATHS );
+        pkgtypes = new ConcurrentHashMap<>(  );
+        pkgtypes.put( MVN_SP_PATH_SET.getPackageType(), MVN_SP_PATH_SET );
     }
 
     @Override
-    public synchronized void registerSpecialPathInfo( SpecialPathInfo pathInfo )
+    public void registerSpecialPathInfo( SpecialPathInfo pathInfo )
     {
-        specialPaths.add( pathInfo );
+        //FIXME: do we need a flag of "isStandard" to add into stdSpecialPaths?
+        registerSpecialPathInfo( pathInfo, PKG_TYPE_MAVEN );
+    }
+
+    @Override
+    public void registerSpecialPathInfo( SpecialPathInfo pathInfo, final String pkgType )
+    {
+        for ( Map.Entry<String, SpecialPathSet> entry : pkgtypes.entrySet() )
+        {
+            if ( pkgType.equals( entry.getKey() ) )
+            {
+                entry.getValue().registerSpecialPathInfo( pathInfo );
+            }
+        }
     }
 
     @Override
     public synchronized void deregisterSpecialPathInfo( SpecialPathInfo pathInfo )
     {
-        specialPaths.remove( pathInfo );
+        //FIXME: do we need a flag of "isStandard" to remove from stdSpecialPaths?
+        deregisterSpecialPathInfo( pathInfo, PKG_TYPE_MAVEN );
```

Comment:
Again, I think this should evaluate against the standard special path info, not the stuff for the maven package